src/main/java/games/strategy/triplea/ai/proAI/ProPurchaseAI.java

```diff
@@ -121,6 +122,100 @@ int repair(int PUsRemaining, final IPurchaseDelegate purchaseDelegate, final Gam
     return PUsRemaining;
   }
 
+  /**
+   * Default settings for bidding:
+   * 1) Limit one bid unit in a territory or sea zone (until set in all territories then 2, etc)
+   * 2) The nation placing a unit in a territory or sea zone must have started with a unit in said territory or sea zone
+   * prior to placing the bid
+   */
+  Map<Territory, ProPurchaseTerritory> bid(final int pus, final IPurchaseDelegate purchaseDelegate,
+      final GameData startOfTurnData) {
+
+    // Current data fields
+    data = ProData.getData();
+    this.startOfTurnData = startOfTurnData;
+    player = ProData.getPlayer();
+    resourceTracker = new ProResourceTracker(pus, data);
+    territoryManager = new ProTerritoryManager(calc);
+    isBid = true;
+    final ProPurchaseOptionMap purchaseOptions = ProData.purchaseOptions;
+
+    ProLogger.info("Starting bid phase with resources: " + resourceTracker);
+    if (!player.getUnits().getUnits().isEmpty()) {
+      ProLogger.info("Starting bid phase with unplaced units=" + player.getUnits().getUnits());
+    }
+
+    // Find all purchase/place territories
+    final Map<Territory, ProPurchaseTerritory> purchaseTerritories = ProPurchaseUtils.findBidTerritories(player);
+
+    int previousNumUnits = 0;
+    while (true) {
+
+      // Determine max enemy attack units and current allied defenders
+      territoryManager.populateEnemyAttackOptions(new ArrayList<>(), new ArrayList<>(purchaseTerritories.keySet()));
+      findDefendersInPlaceTerritories(purchaseTerritories);
+
+      // Prioritize land territories that need defended and purchase additional defenders
+      final List<ProPlaceTerritory> needToDefendLandTerritories =
+          prioritizeTerritoriesToDefend(purchaseTerritories, true);
+      purchaseDefenders(purchaseTerritories, needToDefendLandTerritories, purchaseOptions.getLandFodderOptions(),
+          purchaseOptions.getAirOptions(), true);
+
+      // Find strategic value for each territory
+      ProLogger.info("Find strategic value for place territories");
+      final Map<Territory, Double> territoryValueMap =
+          ProTerritoryValueUtils.findTerritoryValues(player, new ArrayList<>(), new ArrayList<>());
+      for (final Territory t : purchaseTerritories.keySet()) {
+        for (final ProPlaceTerritory ppt : purchaseTerritories.get(t).getCanPlaceTerritories()) {
+          ppt.setStrategicValue(territoryValueMap.get(ppt.getTerritory()));
+          ProLogger.debug(ppt.getTerritory() + ", strategicValue=" + territoryValueMap.get(ppt.getTerritory()));
+        }
+      }
```

Comment:
This could potentialle be replaced with a lambda using the stream api:
```
purchaseTerritories.keySet().forEach(t -> purchaseTerritories.get(t).getCanPlaceTerritories().forEach(ppt -> {
          ppt.setStrategicValue(territoryValueMap.get(ppt.getTerritory()));
          ProLogger.debug(ppt.getTerritory() + ", strategicValue=" + territoryValueMap.get(ppt.getTerritory()));
}));
```